### PR TITLE
fix(di): add better type information to injector.get()

### PR DESCRIPTION
**Pull request by mhevery**
fix(di): add better type information to injector.get()

BREAKING CHANGE:

- `injector.get()` is now parameterize and returns a type based on args.
  - `injector.get(Foo)` returns an instance of `Foo`
  - `injector.get(new OpaqueToken<Foo>(‘Foo’))` returns an instance of 
    `Foo`.
- OpaqueToken is now parameterized.

Migration
- If you declare your own `OpaqueToken`, it must now be parameterized.
- In the past it was possible that the injector token could return a 
  type other than itself. For example 
  `var cars: Car[] = injector.get(Car)` because the provider for `Car` 
  was declared `multi: true`. This is now no longer possible and it is 
  considered a bad form. Proper way to do this is through
  `var CARS = new OpaqueToken<Car[]>;` and then 
  `var cars = injector.get(CARS)`.
- `var foo: any = injector.get(‘something’)` is still supported and 
  results in `any` type, but it’s use is strongly discouraged.

### CODE REVIEW

File: modules/@angular/compiler/src/url_resolver.ts [hunk 1]
```diff
@@ -6,12 +6,13 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
-import {Inject, PACKAGE_ROOT_URL} from '@angular/core';
+import {Inject, OpaqueToken, PACKAGE_ROOT_URL} from '@angular/core';
```
Comment (vicb) [inline modules/@angular/compiler/src/url_resolver.ts]: Why ?

File: modules/@angular/core/src/di/opaque_token.ts [hunk 2]
```diff
@@ -28,7 +28,7 @@
  * error messages.
  * @stable
  */
-export class OpaqueToken {
+export class OpaqueToken<T> {
```
Comment (vicb) [inline modules/@angular/core/src/di/opaque_token.ts]: add docs for T

File: modules/@angular/core/src/application_tokens.ts [hunk 3]
```diff
@@ -53,10 +54,11 @@ export const PLATFORM_INITIALIZER: any = new OpaqueToken('Platform Initializer')
  *
  * @experimental
  */
-export const APP_BOOTSTRAP_LISTENER = new OpaqueToken('appBootstrapListener');
+export const APP_BOOTSTRAP_LISTENER =
+    new OpaqueToken<((compRef: ComponentRef<any>) => void)[]>('appBootstrapListener');
```
Comment (vicb) [inline modules/@angular/core/src/application_tokens.ts]: should we use `_, __, ___, ...` for parameters at this location ?

File: modules/@angular/forms/src/validators.ts [hunk 4]
```diff
@@ -28,7 +29,7 @@ function isEmptyInputValue(value: any) {
  * {@example core/forms/ts/ng_validators/ng_validators.ts region='ng_validators'}
  * @stable
  */
-export const NG_VALIDATORS: OpaqueToken = new OpaqueToken('NgValidators');
+export const NG_VALIDATORS = new OpaqueToken<Array<Validator|Function>>('NgValidators');
```
Comment (vicb) [inline modules/@angular/forms/src/validators.ts]: Array<> => []
should be a tslint rule probably

File: modules/@angular/http/test/http_spec.ts [hunk 5]
```diff
@@ -42,8 +42,8 @@ export function main() {
 
          http = injector.get(Http);
          jsonp = injector.get(Jsonp);
-         jsonpBackend = injector.get(JSONPBackend);
-         xhrBackend = injector.get(XHRBackend);
+         jsonpBackend = injector.get(JSONPBackend) as MockBackend;
```
Comment (vicb) [inline modules/@angular/http/test/http_spec.ts]: why is this needed ?

File: modules/@angular/compiler/src/url_resolver.ts (hunk 1 above)
Comment (mhevery) [inline modules/@angular/compiler/src/url_resolver.ts]: Because the `.d.ts` file now has `PACKAGE_ROOT_URL` from `any` to `OpaqueToken`. So the import is needed to make `.d.ts` happy, even thought `OpaqueToken` is not used in this file directly (only through inference)

File: modules/@angular/core/src/application_tokens.ts (hunk 3 above)
Comment (mhevery) [inline modules/@angular/core/src/application_tokens.ts]: I think names should have a meaning. Would like to keep as is.

File: modules/@angular/core/src/di/opaque_token.ts (hunk 2 above)
Comment (mhevery) [inline modules/@angular/core/src/di/opaque_token.ts]: good point, added.

File: modules/@angular/forms/src/validators.ts (hunk 4 above)
Comment (mhevery) [inline modules/@angular/forms/src/validators.ts]: But that would be `(Validator|Function)[]` which to me looks strange (@chuckjaz agrees).

File: modules/@angular/http/test/http_spec.ts (hunk 5 above)
Comment (mhevery) [inline modules/@angular/http/test/http_spec.ts]: because `jsonBackend` is declared as `MockBackend`. `injector.get(JSONPBackend)` returns type `JSONPBackend` which is a subtype of `MockBackend ` hence cast is required.

File: modules/@angular/core/src/application_tokens.ts [hunk 6]
```diff
@@ -43,7 +44,7 @@ function _randomChar(): string {
  * A function that will be executed when a platform is initialized.
  * @experimental
  */
-export const PLATFORM_INITIALIZER: any = new OpaqueToken('Platform Initializer');
+export const PLATFORM_INITIALIZER = new OpaqueToken<(() => void)[]>('Platform Initializer');
```
Comment (chuckjaz) [inline modules/@angular/core/src/application_tokens.ts]: Consider changing the type to `OpaqueToken<Array<() => void>>`.

File: modules/@angular/core/src/application_tokens.ts (hunk 3 above)
Comment (chuckjaz) [inline modules/@angular/core/src/application_tokens.ts]: Consider changing the type to `OpaqueToken<Array<(compRef: ComponentRef<any>) => void>>`.

**Comment by IgorMinar** [comment]
As raised by several people, this looks like a breaking change that would affect a big part of our ecosystem. Do we really understand the impact? Can we chat about this one before it is merged?

**Review by IgorMinar — CHANGES_REQUESTED** [review]
As raised by others this looks like a bigger breaking change that would have a significant impact on the ecosystem. Our promise was that we are not going to break existing components on npm.

Can we evaluate and document the impact of this change on the ecosystem (existing 3rd party components, applications, tooling, etc) before we merge this in?

File: modules/@angular/core/src/application_tokens.ts (hunk 6 above)
Comment (mhevery) [inline modules/@angular/core/src/application_tokens.ts]: fixed

File: modules/@angular/core/src/application_tokens.ts (hunk 3 above)
Comment (mhevery) [inline modules/@angular/core/src/application_tokens.ts]: fixed

**Comment by mhevery** [comment]
@IgorMinar / @vicb Could you have a look? This should be a lot less breaking now.

File: modules/@angular/benchpress/index.ts [hunk 7]
```diff
@@ -9,7 +9,7 @@
 // Must be imported first, because angular2 decorators throws on load.
 import 'reflect-metadata';
 
-export {Injector, OpaqueToken, Provider, ReflectiveInjector} from '@angular/core';
+export {InjectionToken, Injector, Provider, ReflectiveInjector} from '@angular/core';
```
Comment (vicb) [inline modules/@angular/benchpress/index.ts]: Why not `Token` (or `InjectionProvider`, ...)

File: modules/@angular/compiler/src/aot/static_reflector.ts [hunk 8]
```diff
@@ -507,8 +507,8 @@ export class StaticReflector implements ReflectorReader {
                 staticSymbol = simplifyInContext(context, expression['expression'], depth + 1);
                 if (staticSymbol instanceof StaticSymbol) {
                   if (staticSymbol === self.opaqueToken) {
```
Comment (vicb) [inline modules/@angular/compiler/src/aot/static_reflector.ts]: rename `opaqueToken` ? (grep the whole project)

File: modules/@angular/core/src/di/injection_token.ts [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+/**
+ * Creates a token that can be used in a DI Provider.
+ *
+ * ### Example ([live demo](http://plnkr.co/edit/Ys9ezXpj2Mnoy3Uc8KBp?p=preview))
+ *
+ * ```typescript
+ * var t = new OpaqueToken("value");
```
Comment (vicb) [inline modules/@angular/core/src/di/injection_token.ts]: update

File: modules/@angular/core/src/di/injection_token.ts [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+ *   {provide: t, useValue: "bindingValue"}
+ * ]);
+ *
+ * expect(injector.get(t)).toEqual("bindingValue");
+ * ```
+ *
+ * Using an `OpaqueToken` is preferable to using strings as tokens because of possible collisions
+ * caused by multiple providers using the same string as two different tokens.
+ *
+ * Using an `OpaqueToken` is preferable to using an `Object` as tokens because it provides better
```
Comment (vicb) [inline modules/@angular/core/src/di/injection_token.ts]: ditto * 2

File: modules/@angular/core/src/di/injection_token.ts [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+ * ### Example ([live demo](http://plnkr.co/edit/Ys9ezXpj2Mnoy3Uc8KBp?p=preview))
+ *
+ * ```typescript
+ * var t = new OpaqueToken("value");
+ *
+ * var injector = Injector.resolveAndCreate([
+ *   {provide: t, useValue: "bindingValue"}
+ * ]);
+ *
+ * expect(injector.get(t)).toEqual("bindingValue");
```
Comment (vicb) [inline modules/@angular/core/src/di/injection_token.ts]: Do we still use inline examples ? @IgorMinar

File: modules/@angular/core/src/di/injection_token.ts [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+ * var t = new InjectionToken<string>("value");
+ *
+ * var injector = Injector.resolveAndCreate([
+ *   {provide: t, useValue: "bindingValue"}
+ * ]);
+ *
+ * expect(injector.get(t)).toEqual("bindingValue");
+ * ```
+ *
+ * Using an `InjectionToken` whenever the type you are injection is not reified (does not have a
```
Comment (vicb) [inline modules/@angular/core/src/di/injection_token.ts]: typo you are injection / formatting

re-organize to avoid duplication with the class comment ?

File: modules/@angular/core/src/di/injection_token.ts [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+ *
+ * ```
+ * var x:SomeType = injector.get(new InjectionToken<SomeType>('SomeToken'));
+ * ```
+ *
+ * @deprecated in v4.0.0. Use `InjectionToken<?>` instead.
+ */
+export class InjectionToken<T> extends OpaqueToken {
+  // This unused property is needed here so that TS can differentiate InjectionToken from
+  // InjectionToken since
```
Comment (vicb) [inline modules/@angular/core/src/di/injection_token.ts]: aren't they the same ;)

File: modules/@angular/core/src/di/injector.ts [hunk 14]
```diff
@@ -52,5 +55,10 @@ export abstract class Injector {
    * Injector.THROW_IF_NOT_FOUND is given
    * - Returns the `notFoundValue` otherwise
    */
+  get<T>(token: Type<T>|InjectionToken<T>, notFoundValue?: T): T;
+  /**
+   * @deprecated from v4.0.0 use Type<T> or InjectToken<T>
```
Comment (vicb) [inline modules/@angular/core/src/di/injector.ts]: typo InjectToken

**Comment by vicb** [comment]
I'd much rather use `di.Token` when there is a need to disambiguate over `InjectorToken` but I guess we already have a big legacy in the code base.
LGTM with changes.

File: modules/@angular/benchpress/index.ts (hunk 7 above)
Comment (mhevery) [inline modules/@angular/benchpress/index.ts]: discussed in person

File: modules/@angular/compiler/src/aot/static_reflector.ts (hunk 8 above)
Comment (mhevery) [inline modules/@angular/compiler/src/aot/static_reflector.ts]: replaced

File: modules/@angular/core/src/di/injection_token.ts (hunk 9 above)
Comment (mhevery) [inline modules/@angular/core/src/di/injection_token.ts]: this is for old `OpaqueToken` should be left as is.

File: modules/@angular/core/src/di/injection_token.ts (hunk 10 above)
Comment (mhevery) [inline modules/@angular/core/src/di/injection_token.ts]: ditto

File: modules/@angular/core/src/di/injection_token.ts (hunk 11 above)
Comment (mhevery) [inline modules/@angular/core/src/di/injection_token.ts]: No, but this is for the old text.

File: modules/@angular/core/src/di/injection_token.ts (hunk 12 above)
Comment (mhevery) [inline modules/@angular/core/src/di/injection_token.ts]: fixed

File: modules/@angular/core/src/di/injection_token.ts (hunk 13 above)
Comment (mhevery) [inline modules/@angular/core/src/di/injection_token.ts]: fixed

File: modules/@angular/core/src/di/injector.ts (hunk 14 above)
Comment (mhevery) [inline modules/@angular/core/src/di/injector.ts]: fixed

**Comment by mhevery** [comment]
Updated with @vicb's comments

File: modules/@angular/core/src/di/injection_token.ts [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+ *
+ * Using an `InjectionToken` whenever the type you are injecting is not reified (does not have a
+ * runtime representation) such as when injecting an interface, callable type, array or
+ * parametrized type.
+ *
+ * `InjectionToken` is parametrize on `T` which is the type of object which will be returned by the
+ * `Injector`. This provides additional level of type safety.
+ *
+ * ```
+ * var x:SomeType = injector.get(new InjectionToken<SomeType>('SomeToken'));
```
Comment (vicb) [inline modules/@angular/core/src/di/injection_token.ts]: - Use an interface ?
- `get(new InjectionToken<SomeType>('SomeToken'))` is probably misleading ?

File: modules/@angular/core/src/di/injection_token.ts [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+ * var t = new InjectionToken<string>("value");
+ *
+ * var injector = Injector.resolveAndCreate([
+ *   {provide: t, useValue: "bindingValue"}
+ * ]);
+ *
+ * expect(injector.get(t)).toEqual("bindingValue");
+ * ```
+ *
+ * Using an `InjectionToken` whenever the type you are injecting is not reified (does not have a
```
Comment (vicb) [inline modules/@angular/core/src/di/injection_token.ts]: Using -> use

File: modules/@angular/core/src/di/injection_token.ts [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+export class OpaqueToken {
+  constructor(protected _desc: string) {}
+
+  toString(): string { return `Token ${this._desc}`; }
+}
+
+/**
+ * Creates a token that can be used in a DI Provider.
+ *
+ * ### Example ([live demo](http://plnkr.co/edit/Ys9ezXpj2Mnoy3Uc8KBp?p=preview))
```
Comment (vicb) [inline modules/@angular/core/src/di/injection_token.ts]: remove the plnkr

File: modules/@angular/compiler/src/jit/compiler_factory.ts [hunk 18]
```diff
@@ -40,7 +40,7 @@ const _NO_RESOURCE_LOADER: ResourceLoader = {
           `No ResourceLoader implementation has been provided. Can't read the url "${url}"`);}
 };
 
-const baseHtmlParser = new OpaqueToken('HtmlParser');
+const baseHtmlParser = new InjectionToken('HtmlParser');
```
Comment (vicb) [inline modules/@angular/compiler/src/jit/compiler_factory.ts]: `< HtmlParser >`

File: modules/@angular/compiler/src/metadata_resolver.ts [hunk 19]
```diff
@@ -27,7 +27,7 @@ import {getUrlScheme} from './url_resolver';
 import {MODULE_SUFFIX, SyntaxError, ValueTransformer, visitValue} from './util';
 
 export type ErrorCollector = (error: any, type?: any) => void;
-export const ERROR_COLLECTOR_TOKEN = new OpaqueToken('ErrorCollector');
+export const ERROR_COLLECTOR_TOKEN = new InjectionToken('ErrorCollector');
```
Comment (vicb) [inline modules/@angular/compiler/src/metadata_resolver.ts]: `< ErrorCollector >`

File: modules/@angular/compiler/src/template_parser/template_parser.ts [hunk 20]
```diff
@@ -67,7 +67,7 @@ const TEXT_CSS_SELECTOR = CssSelector.parse('*')[0];
  *
  * This is currently an internal-only feature and not meant for general use.
  */
-export const TEMPLATE_TRANSFORMS = new OpaqueToken('TemplateTransforms');
+export const TEMPLATE_TRANSFORMS = new InjectionToken('TemplateTransforms');
```
Comment (vicb) [inline modules/@angular/compiler/src/template_parser/template_parser.ts]: +type

File: modules/@angular/core/src/application_tokens.ts [hunk 21]
```diff
@@ -19,7 +20,7 @@ import {OpaqueToken} from './di';
  * using this token.
  * @experimental
  */
-export const APP_ID: any = new OpaqueToken('AppId');
+export const APP_ID: any = new InjectionToken('AppId');
```
Comment (vicb) [inline modules/@angular/core/src/application_tokens.ts]: string

File: modules/@angular/core/src/di/injection_token.ts [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+ * `Injector`. This provides additional level of type safety.
+ *
+ * ```
+ * var x:SomeType = injector.get(new InjectionToken<SomeType>('SomeToken'));
+ * ```
+ */
+export class InjectionToken<T> extends OpaqueToken {
+  // This unused property is needed here so that TS can differentiate InjectionToken from
+  // OpaqueToken since otherwise they would have the same shape and be treated as equivalent.
+  private _differentiate_from_OpaqueToken_structurally: any;
```
Comment (vicb) [inline modules/@angular/core/src/di/injection_token.ts]: would it make sense to make it public ?

File: modules/@angular/core/src/di/injector.ts [hunk 23]
```diff
@@ -52,5 +55,10 @@ export abstract class Injector {
    * Injector.THROW_IF_NOT_FOUND is given
    * - Returns the `notFoundValue` otherwise
    */
+  get<T>(token: Type<T>|InjectionToken<T>, notFoundValue?: T): T;
```
Comment (vicb) [inline modules/@angular/core/src/di/injector.ts]: `notFoundValue?: T|null`

File: modules/@angular/core/src/di/injection_token.ts (hunk 15 above)
Comment (mhevery) [inline modules/@angular/core/src/di/injection_token.ts]: made more clear

File: modules/@angular/core/src/di/injection_token.ts (hunk 16 above)
Comment (mhevery) [inline modules/@angular/core/src/di/injection_token.ts]: updated

File: modules/@angular/core/src/di/injection_token.ts (hunk 17 above)
Comment (mhevery) [inline modules/@angular/core/src/di/injection_token.ts]: This refers to the old API docs, which I am not changing.

File: modules/@angular/compiler/src/jit/compiler_factory.ts (hunk 18 above)
Comment (mhevery) [inline modules/@angular/compiler/src/jit/compiler_factory.ts]: There is several hundred of these in our code base. I started doing them, but then I realized that I could not determine the type for most of them so I gave up, and I am not planning on doing them.

File: modules/@angular/compiler/src/metadata_resolver.ts (hunk 19 above)
Comment (mhevery) [inline modules/@angular/compiler/src/metadata_resolver.ts]: Same heere

File: modules/@angular/core/src/di/injection_token.ts (hunk 22 above)
Comment (mhevery) [inline modules/@angular/core/src/di/injection_token.ts]: It has no value other then for TypeScript to disambiguate the types. There is no need to read or write it.

File: modules/@angular/core/src/di/injector.ts (hunk 23 above)
Comment (mhevery) [inline modules/@angular/core/src/di/injector.ts]: `null` can't be used in type position until we upgrade to TS 2.0

File: modules/@angular/core/src/di/injection_token.ts (hunk 11 above)
Comment (IgorMinar) [inline modules/@angular/core/src/di/injection_token.ts]: @vicb I don't care that much about these inline examples right now. We need to do an audit of the whole repo and migrate/delete all of them in one go. I have it on my list of A4. Can happen during RC as this is just a doc improvement.

File: modules/@angular/core/src/di/injection_token.ts [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+ *
+ * expect(injector.get(t)).toEqual("bindingValue");
+ * ```
+ *
+ * Using an `OpaqueToken` is preferable to using strings as tokens because of possible collisions
+ * caused by multiple providers using the same string as two different tokens.
+ *
+ * Using an `OpaqueToken` is preferable to using an `Object` as tokens because it provides better
+ * error messages.
+ * @deprecated since v4.0.0 use `InjectionToken<?>` instead.
```
Comment (IgorMinar) [inline modules/@angular/core/src/di/injection_token.ts]: can you explain why? (can't carry type info)

File: modules/@angular/core/src/di/injection_token.ts [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+ * ### Example ([live demo](http://plnkr.co/edit/Ys9ezXpj2Mnoy3Uc8KBp?p=preview))
+ *
+ * ```typescript
+ * var t = new InjectionToken<string>("value");
+ *
+ * var injector = Injector.resolveAndCreate([
+ *   {provide: t, useValue: "bindingValue"}
+ * ]);
+ *
+ * expect(injector.get(t)).toEqual("bindingValue");
```
Comment (IgorMinar) [inline modules/@angular/core/src/di/injection_token.ts]: please create this as a standalone example and then use @example to include it here. we've done this for other apis already. Please no more new inline code snippets.

**Review by IgorMinar — APPROVED** [review]
please fix the docs. otherwise LGTM!

File: modules/@angular/core/src/di/injection_token.ts (hunk 24 above)
Comment (mhevery) [inline modules/@angular/core/src/di/injection_token.ts]: fixed

File: modules/@angular/core/src/di/injection_token.ts (hunk 25 above)
Comment (mhevery) [inline modules/@angular/core/src/di/injection_token.ts]: done

File: modules/@angular/core/src/di/injection_token.ts (hunk 22 above)
Comment (vicb) [inline modules/@angular/core/src/di/injection_token.ts]: was thinking of extending
